### DISCUSSION TITLE
React Native Connectors and FAQ

### DISCUSSION THREAD

**Discussion by thiagobrez** [wevm/wagmi#3099] [Q&A]
Hey team!

First of all, thanks for the hard work put in here - wagmi is amazing!

I wanted to talk about Wagmi + React Native. I already know of the [React Native support](https://github.com/wagmi-dev/wagmi/discussions/533) discussion here, but my point is slightly different.

Since wagmi is one of the highest-level developer-facing libraries, developers will usually come here first asking for React Native support, hence the issues filed here mentioning errors coming from missing `viem` polyfills, bugs in the WalletConnect repo, and others that are not issues with wagmi itself.

Also, we know that wagmi does not depend on browser APIs and should work well in RN, and it does! All we need is good connectors wrapping mobile SDKs.

So I'd like to propose 2 things, to which I could be sending PRs:
- Add a React Native section in the FAQ page, mentioning what is needed to make sure it will work well (`viem` TextEncoder polyfill and a custom Connector wrapping the mobile SDK of choice)
- Add React Native connectors, such as one wrapping the MetaMask Mobile SDK, instead of extending the InjectedConnector

Just wanted to ask first if the team has any plans for React Native connectors, and if you are accepting PRs in this area, to avoid duplicated work. I have a sample implementation of a MetaMask Mobile Connector that I could polish and contribute if you think it's a good idea.

I will also start a discussion in the WalletConnect repo because their Ethereum Provider depends on their web modal UI, and if we had a version that did not depend on browser APIs, or even used their own RN modal package, then I believe we could also easily create a WalletConnect Mobile Connector for wagmi.

Thanks, let me know what you think!

**Comment by glitch-txs** [comment]
Hi Thiago, could you share your connector?

I just finished integrating [Wagmi into a React Native app](https://github.com/glitch-txs/react-native-wagmi)

We've created a package that covers the needs of polyfills.

I also had to copy/paste the WalletConnect connector + add very small changes to make it work.

I was getting this error with both WalletConnect and Coinbase connectors

```
Error: Requiring unknown module "1304". If you are sure the module exists, try restarting Metro. You may also want to run `yarn` or `npm install`., js engine: hermes
```

Do you know what it could possibly be? Seems more like a bundler issue somehow as the changes I made weren't that relevant.

If we don't get this working with Wagmi connectors maybe we could propose modularizing them into different packages, this way there could be dedicated React Native connectors as well. Mainly for those that depend on external SDKs.

**Reply by thiagobrez** [reply]
Hey @glitch-txs! That's amazing, thanks for jumping in.

So, if I understand correctly, you're basically initiating the EthereumProvider with `showQrModal: false`, to avoid importing the web modal UI. Then you subscribe to the `display_uri` event to get the QR code uri. From here, we could manually build a UI that allowed the user to select the desired wallet and deeplink to it with the uri, just like WalletConnect does. Right?

This is super cool, and already solves half of the problem! But what do you think if we had an option in EthereumProvider like `showRNQrModal: true`, that would dynamically import WalletConnect's RN modal UI instead of the web one? Then the RN modal would show and the user would be able to enjoy the nice WalletConnect UI and also use wagmi 😎

Something like this:

In the app page:
```tsx
  const { connect } = useConnect({
    connector: new WalletConnectConnector({
      chains: [mainnet],
      options:{ projectId:'...', showRNQrModal: true }
    }),
  })
```

In WalletConnectConnector:
```tsx
      this._provider = await EthereumProvider.init({
        showRNQrModal,
        ...
      })
```

And then In the [initialize() function in EthereumProvider](https://github.com/WalletConnect/walletconnect-monorepo/blob/2a4188ee986c3e26a37241601733bcb92018c580/providers/ethereum-provider/src/EthereumProvider.ts#L510-L546), we import conditionally either the web modal or the RN modal:
```tsx
    let WalletConnectModalClass;

    // showQrModal shows the web modal UI
    if (this.rpc.showQrModal) {
      try {
        const { WalletConnectModal } = await import("@walletconnect/modal");
        WalletConnectModalClass = WalletConnectModal;
      } catch {
        throw new Error("To use QR modal, please install @walletconnect/modal package");
      }
    }
    
    // showRNQrModal shows the React Native modal UI
    if (this.rpc.showRNQrModal) {
      try {
        const { WalletConnectModal } = await import("@walletconnect/modal-react-native");
        WalletConnectModalClass = WalletConnectModal;
      } catch {
        throw new Error("To use QR modal, please install @walletconnect/modal package");
      }
    }
    
    if (WalletConnectModalClass) {
        try {
          this.modal = new WalletConnectModalClass({
            ...
          });
        } catch (e) {
          this.signer.logger.error(e);
          throw new Error("Could not generate WalletConnectModal Instance");
        }
      }
    }
```

**Reply by thiagobrez** [reply]
Answering your other questions:

>I also had to copy/paste the WalletConnect connector + add very small changes to make it work.

Cool, looks good. Regarding removing the `#` declaration for private properties, if you prefer I think this could easily be solved by using `['@babel/plugin-transform-private-methods', {loose: true}]`

>I was getting this error with both WalletConnect and Coinbase connectors

I can't reproduce that one, but I see a weird thing in `package.json`: seems like you installed `npm` and `install` by accident. Could be related to that

```
    "install": "^0.13.0",
    "npm": "^10.2.0",
```

Also, wagmi connects succesfully, but I see this other errors in the console as soon as I press the connect button:

```
 ERROR  {"context": "core"} {"context": "core/crypto"} Failed to decode message from topic: '4e8baaeb415cbc8a65dfbe504d76dc8cebe9f78879836a998f28b1658a3de75f', clientId: 'did:key:z6MkkKssAL8h8TXSaYUVCmxnzdK6VV2EdErq8Ho79XpQ9XDi'
 ERROR  {"context": "core"} {"context": "core/crypto"} [Error: No matching key. keychain: 4e8baaeb415cbc8a65dfbe504d76dc8cebe9f78879836a998f28b1658a3de75f]
 ERROR  {"context": "core"} {"context": "core/crypto"} Failed to decode message from topic: '4e8baaeb415cbc8a65dfbe504d76dc8cebe9f78879836a998f28b1658a3de75f', clientId: 'did:key:z6MkkKssAL8h8TXSaYUVCmxnzdK6VV2EdErq8Ho79XpQ9XDi'
 ERROR  {"context": "core"} {"context": "core/crypto"} [Error: No matching key. keychain: 4e8baaeb415cbc8a65dfbe504d76dc8cebe9f78879836a998f28b1658a3de75f]
 ERROR  {"context": "client"} [Error: Missing or invalid. Decoded payload on topic 4e8baaeb415cbc8a65dfbe504d76dc8cebe9f78879836a998f28b1658a3de75f is not identifiable as a JSON-RPC request or a response.]
```

**Reply by glitch-txs** [reply]
> I can't reproduce that one

Have you tried with Wagmi's connectors? not custom ones. Anyways for Coinbase they have a React Native specific SDK so it needs to be rebuild regarless

**Reply by glitch-txs** [reply]
If everything goes as expected we will move forward integrating Web3Modal SDK + Wagmi for React Native.

**Reply by thiagobrez** [reply]
Ah, tried with the custom connector you got there. Ok, I'll try with wagmi's  tomorrow morning and get back to you

>If everything goes as expected we will move forward integrating Web3Modal SDK + Wagmi for React Native.

Nice! let me know if I can help

**Reply by donkey-donkey** [reply]
we are getting an error like this in react native with coinbase wallet when we specify 
connectors in the config
```

Metro error: Requiring unknown module "undefined". If you are sure the module exists, try restarting Metro. You may also want to run `yarn` or `npm install`.

(pp/node_modules/@metamask/sdk/dist/node/cjs/metamask-sdk.js:49:7069)
```

**Reply by glitch-txs** [reply]
It's better to recreate the connectors yourself, at WalletConnect we did that for wagmi v1
Also the Coinbase connector from Wagmi is for web. Coinbase has a specific SDK for React Native.

**Comment by aditya-aarna-lab** [comment]
Hi
Im using walletconnect + wagmi in my react-native app. However for metamask im facing a lot of issues. 
1. Auto redirect does not work
2. Most of the time after connection, i don't get the provider
3. get errors like "No matching key. pairing {{id}}..."

I have a few concerns
1. Im now trying to integrate Metamask directly to my app. but I want to keep using wagmi in my react native app. Is there any way I can continue using wagmi even if im using metamask-sdk. Most importantly, it this a good approach?
2. do we have any other proficient option?

Could you please help me with any example?
